### ISSUE TITLE
Export script (Node version) - Releasing

### ISSUE THREAD

**Issue by didoo** [hashicorp/flight#188]
#### Description
This is the last part (the "releasing") of the new Node.js version of the "Figma to Flight" export script.
What it does is to take the bundles generated in the "build" step in different formats, prepare them as NPM packages and release them to the registry.

The Ember Addon instead will be released following the standard process in Ember (but we'll try to find a way to automate also this release).

#### Expected logic
This is an outline of what we intend to implement

```
- clone (or just process, if we don't commit it to the repo)) the `dist` folder that was generated by the "build" step
- bump the version of the main package.json file and all the derived ones [notice: TBD the versioning strategy]
- prepare the "bundles" folders to be released as NPM packages
- release each target package to NPM
  - TBD if we need to check that the use has already done a login and has the right permissions
- cleanup any temporary file we created in the process
```
#### Things to consider/keep in mind

- We have discussed the versioning strategy (see #154) and we have agreed that the version of the packages and the version of the Ember addon will be independent.
- We still have to understand if it's possible to run a `yarn release` (or similar) command via Node script, or needs to be done via CLI (in which case the "release" process may change a little).

#### Tools & Links
- Yarn Workspaces - https://classic.yarnpkg.com/en/docs/workspaces/
- Lerna - https://lerna.js.org/
- Lerna Wizard - https://github.com/webuniverseio/lerna-wizard
- Release-it - https://github.com/release-it/release-it (see #115)
- a couple of examples in which Lerna is used with monorepo+Ember packages: [example 1](https://github.com/dunkinbase/ember-elements) - [example 2](https://github.com/freshworks/nucleus)
- https://github.community/t/multiple-packages-from-the-same-repo-root-level-package-json-needed/17691
- npm-publish-all (package) - https://www.npmjs.com/package/npm-publish-all
- https://medium.com/ah-technology/a-guide-through-the-wild-wild-west-of-setting-up-a-mono-repo-with-typescript-lerna-and-yarn-ed6a1e5467a
- how to install a local node package manager https://github.com/lerna/lerna/issues/51#issuecomment-348256663 (plus thread about how to do a dry-run of `lerna publish`)

#### Tasks
- [x] Pair with @amyrlam and understand what options we have, what is the expected content of the bundles to be ready to be published, and what we can do to automate this process as much as possible
- [x] Check how we can release multiple packages with a single command
- [x] Setup a process that takes the bundles and release them on the registry
- [x] Update the README file with the documentation for the "release" step

**Comment by didoo** [comment]
After a pair session with @amyrlam in which we have discussed what possible strategies to adopt for the "release" step of the new Node export script, we have agreed:

- to simplify the folder structure and location of the package.json files for the `flight-icons` folder: now the `dist` folder will contain only the SVG files in a `svg` folder and the `catalog.json` file; if in the future we will need to release the assets also in other formats (eg. CSS or React) we will change the script(s) accordingly (and pay the price then).
- to test different technical options: I will test Lerna (and simple NPM scripts) and @amyrlam will test [release-it](https://github.com/release-it/release-it).

On my side, I've done *a lot* of testing to try to figure out what possible solutions were, below a transcript of what I wanted to do and what I learned in the process (for future reference).

------

# Current situation

At the moment, publishing the packages for the `flight-icons` (standalone SVGs) and `ember-flight-icons` (Ember addon) requires a set of manual steps:

- cd in the `flight-icons` folder
- bump the semver version (manually in the package.json)
- run `npm publish`
- cd in the `ember-flight-icons` folder
- bump the semver version (manually in the package.json)
- run `npm publish`

the versioning strategy needs to be independent (as we have agreed) which means depending on what changed, we have to do things slightly diffently:

- if what changed is an icon (in Figma, exported in the repo) then we need to bump both `flight-icons` and `ember-flight-icons`
- if what changed is something in the Ember addon (eg. the <FlightIcon> code) then we need to bump *only* `ember-flight-icons`

which means there is an asymmetry between the two.

# What I would like to achieve

Because of all these implicit knowledge (that yes, we can document somewhere, but still is quite human error prone) I would prefer to automate as much as possible the process.

Ideally I would like to have:
- a `release` command in the `flight-icons` folder that does:
  - prepare the `dist` folder for release as NPM packages (eg. copying there the relevant files like README, package.json, etc)
  - prompt the user about what _semver_ version to use for the bump
  - automatically bump the version in the `package.json` file
  - publish the `dist` folder automatically on the NPM registry
  - cleanup the `dist` folder removing temporary files
  - (optional) when the Ember addon will consume the `flight-icons` as package/dependency, automatically update the version in the `package.json` file in the `ember-flight-icons` folder
- a `release` command in the `ember-flight-icons` folder that does:
  - the standard release for an Ember addon

In this way, we can independently publish both packages, but with a high level of automation.

# Setup of a local NPM registry for testing

To test the release of packages without actually polluting the real/production NPMjs repository, I have setup a local registry using [Verdaccio](https://verdaccio.org/docs/what-is-verdaccio) which is an open source solution, very easy to setup and use.

You can follow [the instructions here](https://verdaccio.org/docs/installation) but essentially what you have to do is:

- install the package: `npm install -g verdaccio` (this will install it globally)
- launch the service: `verdaccio`
- this will serve a web frontend to the registry at the URL http://localhost:4873/
- add a user to the registry: `npm adduser --registry http://localhost:4873` (this will ask you a username/password/email, I suggest to use test/test/test@test.com because is just a local instance); this will also authenticate you with the registry so when you publish yuo don't need to login.

Now you need to add this entry in the `package.json` files of both `flight-icons` and `ember-flight-icons`:

```
"publishConfig": {
    "registry": "http://localhost:4873"
},
```

This will make sure the packages are published on Verdaccio. Once a package is published, the web frontend accesible at http://localhost:4873/ will show you all the details about the packages (and you can also download the tarballs, to check their content).

_Notice: once you've done testing, you can remove verdaccio via `npm uninstall -g verdaccio` and then remove the files he created using `rm -fr ~/.local/share/verdaccio && rm -fr .config/verdaccio`._

# Node scripts

The first attempt was to simply use a combination of Node scripts declared in the `package.json` file to be able to release the `flight-icon` package directly from its `dist` folder.

After many (many!) attempts of trial and error, I came up with this solution that actually works and does what we expect:

```

"scripts": {
    ...
    "release:pre": "cp -f package.json dist/ && cp -f README.md dist/ && cp -f CONTRIBUTING.md dist",
    "release:bump": "./node_modules/.bin/bump package.json \"dist/package.json\"",
    "release:publish": "yarn publish dist/ --new-version $npm_package_version",
    "release:post": "rm dist/package.json && rm dist/README.md && rm dist/CONTRIBUTING.md",
    "release": "yarn release:pre && yarn release:bump && yarn release:publish && yarn release:post"
},
```

This is the content of the `package.json` file in the `flight-icons` folder (the `dist` folder contains *only* the files generated in the build steps, not other relevant files like the READMEs or a clone `package.json` file).

what I am doing here is:
- in pre-release copy the relevant files in the `dist` folder
- run an interactive command "bump" (I've used [version-bump-prompt](https://www.npmjs.com/package/version-bump-prompt]) that ask the user which kind of bump they want (major, minor, patch, specific)
    - this will update both the main `package.json` file as well as the copy in the `dist` folder
- do a `yarn-release` using as folder the `dist` folder and as version the one that has just been updated (the `$npm_package_version` is an environment variable in Node that reads from the `package.json` file)
- in post-release remove the files previously added to the `dist` folder, so it remains cleaner

This works well, and in the future

## Considerations about the NPM scripts

1 - I am not a big fan of using Node scripts:
- they can be very brittle (just move around a folder/file and everything breaks)
- they're very hard to understand and debug (and so also maintain)
- they're based on conventions, implicit/hidden knowledge, you need to know exactly what a command does (you can't inspect its code)
- there's no error recovery and very limited logging
so my first choice is always to go with code-based Node scripts.

2 - All of these steps (expecially the pre/post ones) could be converted to custom Node.js scripts (like the "sync" and "build" ones); this would give us more flexibility and control, plus would be easier to extend it in the future in case we need to change the version number in the `package.json` file of the Ember addon.

3 - I have tried to use run the `yarn release` directly from the `flight-icons` folder, using `.npmignore` and the special `files` entry in the `package.json` file (`"files": ["dist"]`) but this doesn't work: yarn releases the entire `dist` folder in the tarball, not its contents as is said is some (probably outdated) web pages.

#  Lerna

The second attempt was to use [Lerna](https://lerna.js.org/), a tool used to manage monorepos with multiple packages (so, our situation).

Configuring Lerna on our project was not as difficult as I was expecting (I have also looked at these two project as examples, [project 1](https://github.com/dunkinbase/ember-elements) / [project 2](https://github.com/freshworks/nucleus)):

Firts of all I installed Lerna in the root of the project: `cd flight && yarn add -D lerna` (not as a global module). This created a `package.json` in the root (it didn't exist), with a minimal entry:

```
{
  "name": "flight",
  "license": "MPL-2.0",
  "devDependencies": {
    "lerna": "^4.0.0"
  }
}
```

Then I launched the initialization of lerna via `yarn lerna init` (see [here for details](https://github.com/lerna/lerna/blob/main/commands/init/README.md)), replied to the prompt (choosing "independent" for the versioning of the packages) and this created a `lerna.json` in the root:

```
{
  "packages": [
    "packages/*"
  ],
  "version": "independent",
  "command": {
    "publish": {
      "conventionalCommits": true,
      "message": "chore(release): publish",
      "allowBranch": "master"
    }
  }
}
```
and I updated it to fit our setup:
- updated the `packages` entry to point to our two distinct packages
- added our Verdaccio local registry
- allowed to publish from every branch, not only master (so I could do it from my development branch)

```
{
  "packages": [
    "flight-icons",
    "ember-addon"
  ],
  "version": "independent",
  "npmClient": "yarn",
  "useWorkspaces": true,
  "command": {
    "publish": {
      "conventionalCommits": true,
      "message": "chore(release): publish",
      "registry": "http://localhost:4873/",
      "allowBranch": "*"
    }
  }
}
```
There was nothing to do on the existing `package.json` files for `flight-icons` and `ember-flight-icons` so I started to play and experiment with the Lerna commands (see [here for details](https://www.nngroup.com/articles/design-team-statistics/)).

What I found is that Lerna is *extremely* opinionated on what is the flow that should be followed to release a package. By default Lerna:
- does a git reset hard on the repo (and you lose the existing changes) → you have to pass a `--no-git-reset` to the command you execute
-- commits a CHANGELOG.md file to the repo (probably because of the `conventionalCommits` option?) → you have to pass `--no-changelog` to avoid it
- commits and push all the changes that happened during the command you executed (so package.json changes, CHANGELOG, etc) → you have to pass `--no-push` to avoid it
- creates a tag in git for the version just created → you have to pass `--no-git-tag-version` to avoid it

For our purpose, the two relevant commands in Lerna are [`version`](https://github.com/lerna/lerna/blob/main/commands/version#readme) and [`publish`](https://github.com/lerna/lerna/blob/main/commands/publish#readme).

The first one is like `publish` but without the actual publishing on a repo. So a tipical command would look like:
```
yarn lerna version minor [--no-push --no-git-tag-version --no-changelog]
```

The second one does actually publish on the (local, in our tests) NPM repo:
```
yarn lerna publish minor [--no-push --no-git-reset --no-git-tag-version --no-changelog]
```

You can also ask Lerna to build *all* your packages before actually publishing (this will make sure your packages are in the rights status before being published!):

```
yarn lerna run build --stream
```

(the stream options triggers concurrency in the builds).

One thing that made me excited was that there is an option `--contents` to tell Lerna to use the `dist` folder for the publishing of the packages. If you run the command

```
yarn lerna publish [version] --contents dist [options]
```

will publish not the packages' folders declared in the `Lerna.json` file but their `dist` subfolders. Only thing is: you can't specify different folders for different packages, and this I discovered is an issue, because the `dist` folder in the `ember-flight-icon` package doesn't contain the bundle that is released as package; in this case, the bundle is the `ember-flight-icon` folder itself, and the content to be published is controlled via the `.npmignore` file (which means, our initial idea to use the `dist` folder for both is now dead).

_Notice: Lerna is smart enough to detect if there is a real change or not in the repo: if you just try to release a new version without actual changes, you will see the message: “Current HEAD is already released, skipping change detection.”_

# Considerations about Lerna

1 - Even if it's simple to setup, I've found it quite hard to test (probably also because of my lack of knowledge, it was the first time I used it myself). The fact that is so opinionated gets in the way to do things as you would expect, you have to get to do it how Learna expects them.

2 - I have done a test to make sure that Lerna doesn't complain if a package is released directly (not via Lerna itself, even though I think this is possible too in Lerna). I have released the packages via Lerna, then release *only* `ember-flight-icon` via `yarn release` and the released again via Lerna, and everything worked perfectly. This means that what described in the initial part, about the need to do independent releasing, is not an issue with Lerna.

3 - Using Lerna will likely need some learning curve for the team: it took me almost a day to start to understand how it works (and I probably understand only the slightest part of it!)

4 - Just so we know, potentially Lerna can take care of only the versioning, and delegate the publishing to something else (eg. someone was mentioning they do the release via `lerna exec -- "npm publish || exit 0"`)

# Final considerations

At the moment both solutions kinda work. I'll wait to hear from @amyrlam her findings and decide together what is the best option for us (and of course @mel @brian your input would be great too).

One thing to consider when we will make the decision is how the different solutions work with GitHub Actions. Personally, I have some doubts about using GH Actions for this kind of process:
- it covers only one scenario (one icon is updated)
- not sure if you can decide which semver version you want to use (to be checked)
- not sure if it also automatically release the packages (to be checked)
One thing I would see it working is in automatically creating a PR ticket for a change, and leave to a human the the review/approval and npm publish step.

In any case, whatever solution we adopt, I intend to change the way we publish the package, and get rid of the `dist` folder as is now.  The idea is to get rid of `src` and `dist`, have `svg-original` for the exported SVGs and `svg` for the processed SVGs, and a single `catalog.json` file, and then use `.npmignore` for selectively publish files in the package.

_Notice: I already know that if in the future we will need to support multiple outputs, like React or CSS, there will be a non trivial refactoring to do, but at the moment it's hard to sell/justify as things are standing, with one single type of output consumed)._

**Comment by Dhaulagiri** [comment]
> we have to do things slightly diffently:
> 
> * if what changed is an icon (in Figma, exported in the repo) then we need to bump both flight-icons and ember-flight-icons
> * if what changed is something in the Ember addon (eg. the code) then we need to bump only ember-flight-icons
> which means there is an asymmetry between the two.

If this is a concern I would propose that we could release a package with the icons and a separate package with the sprite and completely separate the ember addon's direct tie into the icon release process. It would be a consumer of `flight-icons-svg` or whatever and the version number of the addon can progress independent of the icon set itself.

> I would prefer to automate as much as possible the process.

Playing devil's advocate, my impression from reading the rest of the findings is that it might not actually be worth the effort to automate this process at this point given the challenges outlined above and that it seems unlikely that the icon set will require changes on such a regular basis that automation will give us a ton of value _for now_. With proper documentation (something like [this](https://github.com/primer/octicons/blob/79de5ecc37e1ff7ac0893dfa306151bad3458663/CONTRIBUTING.md#releasing-changes) perhaps) the process should be something we can do on an as needed basis and revisit if the process feels too onerous or we are releasing more than I'm assuming we will.

**Comment by MelSumner** [comment]
My first thought was - I didn't think that releasing things would get this complex, but maybe having a design system in a monorepo will require this sort of complexity. That is a specific combination that I have not experienced before now- I am used to a single design but multiple packages.  

Following the logic through, I think if we know we will need things things (soon-ish), it's good to figure them out while we have the time. I would say they are important but not necessarily urgent. I don't know, though, if we have anything particularly urgent, so it seems good to experiment and try things now while things are a little slower. Of course this is just my two cents!

**Comment by didoo** [comment]
> If this is a concern I would propose that we could release a package with the icons and a separate package with the sprite and completely separate the ember addon's direct tie into the icon release process. It would be a consumer of `flight-icons-svg` or whatever and the version number of the addon can progress independent of the icon set itself. 

But that's what I was doing before and was not approved! :) Hence I removed all the recent simplifications of the pipeline.
In any case, the problem persists: you still have to do a *release* of the `ember-flight-icons` bundle when the `flight-icons` version is updated in the `package.json` file. Re. the independent progression, that's not a problem per se.
Update: after discussion, I understand what you mean, using major/minor requirement in the package.json. Mmm, don't think is very resilient, personally I would prefer a more general approach.

> Playing devil's advocate, my impression from reading the rest of the findings is that it might not actually be worth the effort to automate this process at this point given the challenges outlined above and that it seems unlikely that the icon set will require changes on such a regular basis that automation will give us a ton of value _for now_. With proper documentation (something like [this](https://github.com/primer/octicons/blob/79de5ecc37e1ff7ac0893dfa306151bad3458663/CONTRIBUTING.md#releasing-changes) perhaps) the process should be something we can do on an as needed basis and revisit if the process feels too onerous or we are releasing more than I'm assuming we will.

As discussed in the sync meeting, my biggest fear is that having a lot of "as needed" things to do, in a few months we become stuck in dozens of activities we need to do in parallel (often urgently because something suddenly don't work anymore). Personally in my experience this is detrimental to the velocity of a design system, but I see your point so I'll make the minimal changes as agreed (PR coming straight away).
